File: 2025/day01.cpp
// https://adventofcode.com/2025/day/1

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <iostream>
#include <print>
#include <ranges>
#include <span>

fn parse(String const& filename) -> Vec<i32> {
  return aoc::views::read_lines(filename) |
         std::views::transform([](str line) {
           auto distance = aoc::to_number<i32>(line.substr(1));
           if (line[0] == 'L') {
             distance = -distance;
           }
           return distance;
         }) |
         aoc::ranges::to<Vec<i32>>();
}

fn solve_case1(std::span<const i32> sequence) -> i32 {
  auto current = 50;
  return aoc::ranges::accumulate(
      sequence | std::views::transform([&](i32 rotation) {
        // Not fully correct calculation for the modulo,
        // but it identifies zeros correctly
        current = (current + rotation) % 100;
        return static_cast<i32>(current == 0);
      }),
      0);
}

fn solve_case2(std::span<const i32> sequence) -> i32 {
  auto current = 50;
  return aoc::ranges::accumulate(
      sequence | std::views::transform([&](i32 rotation) {
        let current_zero = (current == 0);
        current += rotation;
        let num_zeros = aoc::abs(current / 100) +
                        static_cast<i32>((current <= 0) && !current_zero);
        // Here we have to use the full formula for the modulo
        current = ((current % 100) + 100) % 100;
        return num_zeros;
      }),
      0);
}

int main() {
  std::println("Part 1");
  let example = parse("day01.example");
  AOC_EXPECT_RESULT(3, solve_case1(example));
  let input = parse("day01.input");
  AOC_EXPECT_RESULT(1048, solve_case1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(6, solve_case2(example));
  AOC_EXPECT_RESULT(6498, solve_case2(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2025/src/day01.rs
fn parse(filename: &str) -> Vec<i32> {
    aoc::file::read_lines(filename)
        .iter()
        .map(|line| {
            let mut distance = line[1..].parse::<i32>().unwrap();
            if (line.chars().next().unwrap() == 'L') {
                distance = -distance;
            }
            distance
        })
        .collect::<Vec<_>>()
}

fn solve_case1(sequence: &[i32]) -> i32 {
    let mut current = 50;
    sequence
        .iter()
        .map(|&rotation| {
            // Not fully correct calculation for the modulo,
            // but it identifies zeros correctly
            current = (current + rotation) % 100;
            (current == 0) as i32
        })
        .sum()
}

fn solve_case2(sequence: &[i32]) -> i32 {
    let mut current = 50;
    sequence
        .iter()
        .map(|&rotation| {
            let current_zero = (current == 0);
            current += rotation;
            let num_zeros = (current / 100).abs() + ((current <= 0) && !current_zero) as i32;
            // Here we have to use the full formula for the modulo
            current = current.rem_euclid(100);
            num_zeros
        })
        .sum()
}

fn main() {
    println!("Part 1");
    let example = parse("day01.example");
    assert_eq!(3, solve_case1(&example));
    let input = parse("day01.input");
    assert_eq!(1048, solve_case1(&input));

    println!("Part 2");
    assert_eq!(6, solve_case2(&example));
    assert_eq!(6498, solve_case2(&input));
}
